### ISSUE TITLE
Array printer improvements

### ISSUE THREAD

**Issue by Upabjojr** [sympy/sympy#22939]
Currently, N-dim array objects print as nested lists on a single line.

It would be nice to:
- display the content in multiple line (like matrices do),
- add the class constructor in the printer (so that the printed code can be executed as Python code).

**Comment by SATAKSHI08** [comment]
@Upabjojr I would like to work on this!

**Comment by ThePauliPrinciple** [comment]
I haven't worked on this, but I am not sure if Upabjojr has already started.

**Comment by SATAKSHI08** [comment]
![image](https://user-images.githubusercontent.com/56965113/151599399-269e053a-f6bb-4632-bd4d-7a3046d3d7b4.png)

just confirming that we want the print(a) to be like matrices output?

**Comment by Upabjojr** [comment]
> ![image](https://user-images.githubusercontent.com/56965113/151599399-269e053a-f6bb-4632-bd4d-7a3046d3d7b4.png)
> 
> just confirming that we want the print(a) to be like matrices output?

The one you are showing is the LaTeX printer. That one is already working. I opened this issue for the StrPrinter.

**Comment by Upabjojr** [comment]
> I haven't worked on this, but I am not sure if Upabjojr has already started.

No, I don't really care about the printers.

**Comment by ThePauliPrinciple** [comment]
Matrices print as a single line for me, e.g.:
```python
>>> from sympy import MatrixSymbol
>>> str(MatrixSymbol('M',2,2).as_explicit())
Matrix([[M[0, 0], M[0, 1]], [M[1, 0], M[1, 1]]])
```

**Comment by oscarbenjamin** [comment]
Depends what environment you use. This is isympy:
```python
In [4]: MatrixSymbol("M", 2, 2).as_explicit()
Out[4]: 
⎡M₀₀  M₀₁⎤
⎢        ⎥
⎣M₁₀  M₁₁⎦
```

**Comment by ThePauliPrinciple** [comment]
So which printer(s) should be updated? I think `repr.py` and `str.py` don't do multiline printing for `Matrix`.

**Comment by oscarbenjamin** [comment]
The output I showed is from the unicode pretty printer.

**Comment by ThePauliPrinciple** [comment]
So
```python
>>> from sympy import Matrix
>>> from pprint import pprint
>>> pprint(Matrix([[0,1],[2,3]]))
Matrix([
[0, 1],
[2, 3]])
```
should be mirrored in `Array`.

**Comment by Upabjojr** [comment]
Maybe the real problem is that
```python
eval(str(array)) == array
```
and
```python
type(eval(str(array))) == type(array)
```
are currently not returning `True`.

**Comment by harshrishimiglani** [comment]
@Upabjojr I would like to take up with this issue. I am planning to use matrix function to print it as same.

**Comment by wermos** [comment]
I don't think the built-in `pprint` function from the `pprint` module is supposed to work the way `pprint` from `sympy.printing.pretty` works.
```python
from sympy import *
from sympy.vector import CoordSys3D, Del
from pprint import pprint

init_printing()

delop = Del()
CC_ = CoordSys3D("C")
x,    y,    z    = CC_.x, CC_.y, CC_.z
xhat, yhat, zhat = CC_.i, CC_.j, CC_.k

t = symbols("t")
ten = symbols("10", positive=True)
eps, mu = 4*pi*ten**(-11), ten**(-5)

Bx = 2 * ten**(-4) * cos(ten**5 * t) * sin(ten**(-3) * y)
vecB = Bx * xhat
vecE = (1/eps) * Integral(delop.cross(vecB/mu).doit(), t)

pprint(vecB)
print()
pprint(vecE)

from sympy.printing.pretty import pprint

pprint(vecB)
print()
pprint(vecE)```
gives me
```
(2*sin(C.y/10**3)*cos(10**5*t)/10**4)*C.i

(10**11*Integral(-2*cos(C.y/10**3)*cos(10**5*t)/10**2, t)/(4*pi))*C.k
⎛     ⎛y_C⎞    ⎛  5  ⎞⎞
⎜2⋅sin⎜───⎟ i_C⋅cos⎝10 ⋅t⎠⎟
⎜     ⎜  3⎟           ⎟
⎜     ⎝10 ⎠           ⎟
⎜─────────────────────⎟
⎜           4         ⎟
⎝         10          ⎠

⎛     ⌠                           ⎞
⎜     ⎮       ⎛y_C⎞    ⎛  5  ⎞    ⎟ k_C
⎜     ⎮ -2⋅cos⎜───⎟⋅cos⎝10 ⋅t⎠    ⎟
⎜     ⎮       ⎜  3⎟               ⎟
⎜  11 ⎮       ⎝10 ⎠               ⎟
⎜10  ⋅⎮ ─────────────────────── dt⎟
⎜     ⎮             2             ⎟
⎜     ⎮           10              ⎟
⎜     ⌡                           ⎟
⎜─────────────────────────────────⎟
⎝               4⋅π               ⎠
```
as output.

**Comment by cbm755** [comment]
Re: `eval(str(array)) == array` being True: I don't think that is a SymPy guarantee.

What should work is `eval(srepr(array) == array` (and I believe that one does work already).

I do think `str(array)` should be `Array(...)` just not for this reason.

**Comment by Upabjojr** [comment]
> I do think `str(array)` should be `Array(...)` just not for this reason.

Yes, one can confuse an array with Python's nested lists with the current printer.

**Comment by asmeurer** [comment]
It's not a strong guarantee, but the str printers should generally strive to be copy-pastable.
